# Importing the Data

In [1]:
import re

# to handle datasets
import pandas as pd
import numpy as np

# for visualization
import matplotlib.pyplot as plt
import seaborn as sns

# to divide train and test set
from sklearn.model_selection import train_test_split

# feature scaling
from sklearn.preprocessing import StandardScaler

# to build the models
from sklearn.pipeline import make_pipeline 

# import datetime
import datetime

# encoders
from category_encoders import OrdinalEncoder
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler

# model creation
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline 
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import RandomizedSearchCV 
import pickle

# to visualise al the columns in the dataframe
pd.pandas.set_option('display.max_columns', None)
pd.pandas.set_option('display.max_rows', 100)

In [2]:
# load the csvs - it is available open source and online
df = pd.read_csv('2019_Winter_Data_Science_Intern_Challenge_Data_Set.csv')
df.head()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11

In [3]:
# order id can be dropped since theres only unique values
df['order_id'].value_counts().value_counts()

1    5000
Name: order_id, dtype: int64

# App Question

Question 1: Given some sample data, write a program to answer the following

On Shopify, we have exactly 100 sneaker shops, and each of these shops sells only one model of shoe. We want to do some analysis of the average order value (AOV). When we look at orders data over a 30 day window, we naively calculate an AOV of $3145.13. Given that we know these shops are selling sneakers, a relatively affordable item, something seems wrong with our analysis. 

Think about what could be going wrong with our calculation. Think about a better way to evaluate this data. 
What metric would you report for this dataset?
What is its value?


# Data Exploration

In [4]:
def wrangle(path):
    df = pd.read_csv(path)

    # convert created_at column to a datetime object
    df['created_at']= pd.to_datetime(df['created_at'])

    # sort the dataframe by date
    df = df.sort_values(by=['created_at'])

    # drop high-cardinality and irrelevant features
    df.drop('order_id', axis=1, inplace=True)

    return df

In [5]:
path = '2019_Winter_Data_Science_Intern_Challenge_Data_Set.csv'
df = wrangle(path)
df.head()

shop_id  user_id  order_amount  total_items payment_method  \
1862       39      738           536            4           cash   
1741       39      910           268            2           cash   
3228       97      912           324            2           cash   
1267       80      798           290            2    credit_card   
2689       49      799           258            2    credit_card   

              created_at  
1862 2017-03-01 00:08:09  
1741 2017-03-01 00:10:19  
3228 2017-03-01 00:14:12  
1267 2017-03-01 00:19:31  
2689 2017-03-01 00:22:25

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5000 entries, 1862 to 2457
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   shop_id         5000 non-null   int64         
 1   user_id         5000 non-null   int64         
 2   order_amount    5000 non-null   int64         
 3   total_items     5000 non-null   int64         
 4   payment_method  5000 non-null   object        
 5   created_at      5000 non-null   datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(1)
memory usage: 273.4+ KB


In [7]:
df.tail()

shop_id  user_id  order_amount  total_items payment_method  \
2630       53      940           112            1    credit_card   
1685       34      818           244            2           cash   
1474       21      815           142            1           cash   
317        52      848           292            2           cash   
2457       95      700           168            1    credit_card   

              created_at  
2630 2017-03-30 23:12:13  
1685 2017-03-30 23:16:10  
1474 2017-03-30 23:26:54  
317  2017-03-30 23:41:34  
2457 2017-03-30 23:55:35

In [8]:
df['payment_method'].value_counts()

credit_card    1735
debit          1671
cash           1594
Name: payment_method, dtype: int64

In [9]:
df.sort_values(by=['order_amount'])

shop_id  user_id  order_amount  total_items payment_method  \
4760       92      937            90            1          debit   
1843       92      987            90            1          debit   
2092       92      986            90            1          debit   
4414       92      927            90            1    credit_card   
228        92      757            90            1          debit   
...       ...      ...           ...          ...            ...   
4868       42      607        704000         2000    credit_card   
15         42      607        704000         2000    credit_card   
1602       42      607        704000         2000    credit_card   
2969       42      607        704000         2000    credit_card   
3332       42      607        704000         2000    credit_card   

              created_at  
4760 2017-03-20 07:37:28  
1843 2017-03-06 07:01:09  
2092 2017-03-04 06:44:05  
4414 2017-03-17 09:57:01  
228  2017-03-13 23:57:51  
...                  ...  
4868 2017-03-22 04:00:00  
15   2017-03-07 04:00:00  
1602 2017-03-17 04:00:00  
2969 2017-03-28 04:00:00  
3332 2017-03-24 04:00:00  

[5000 rows x 6 columns]

In [10]:
# any order amount greater than $10000 came from two different stores
df[df['order_amount'] > 10000]

shop_id  user_id  order_amount  total_items payment_method  \
4311       78      960         51450            2          debit   
4646       42      607        704000         2000    credit_card   
520        42      607        704000         2000    credit_card   
4412       78      756         51450            2          debit   
4040       78      852         25725            1           cash   
2821       78      814         51450            2           cash   
60         42      607        704000         2000    credit_card   
2492       78      834        102900            4          debit   
4715       78      818         77175            3          debit   
2297       42      607        704000         2000    credit_card   
15         42      607        704000         2000    credit_card   
511        78      967         51450            2           cash   
4420       78      969         77175            3          debit   
1436       42      607        704000         2000    credit_card   
3780       78      889         25725            1           cash   
2153       42      607        704000         2000    credit_card   
160        78      990         25725            1    credit_card   
3167       78      927         51450            2           cash   
2922       78      740         25725            1          debit   
3705       78      828         51450            2    credit_card   
2270       78      855         25725            1    credit_card   
1362       42      607        704000         2000    credit_card   
1056       78      800         25725            1          debit   
4918       78      823         25725            1           cash   
2906       78      817         77175            3          debit   
3403       78      928         77175            3          debit   
3724       78      766         77175            3    credit_card   
1193       78      944         25725            1          debit   
493        78      983         51450            2           cash   
1602       42      607        704000         2000    credit_card   
2818       78      869         51450            2          debit   
1384       78      867         25725            1           cash   
1452       78      812         25725            1    credit_card   
2548       78      861         25725            1           cash   
1204       78      970         25725            1    credit_card   
4192       78      787         77175            3    credit_card   
617        78      760         51450            2           cash   
3151       78      745         25725            1    credit_card   
2512       78      935         51450            2          debit   
1562       42      607        704000         2000    credit_card   
3440       78      982         25725            1          debit   
4079       78      946         51450            2           cash   
3101       78      855         51450            2    credit_card   
4868       42      607        704000         2000    credit_card   
2690       78      962         77175            3          debit   
4505       78      866         25725            1          debit   
1104       42      607        704000         2000    credit_card   
3332       42      607        704000         2000    credit_card   
2564       78      915         77175            3          debit   
4882       42      607        704000         2000    credit_card   
4584       78      997         25725            1           cash   
3085       78      910         25725            1           cash   
2495       78      707         51450            2           cash   
2773       78      890         25725            1           cash   
490        78      936         51450            2          debit   
1259       78      775         77175            3    credit_card   
2452       78      709         51450            2           cash   
691        78      878        154350            6          deb

# Using Interquartile Range to Find Outliers

In [11]:
df.describe()

shop_id      user_id   order_amount  total_items
count  5000.000000  5000.000000    5000.000000   5000.00000
mean     50.078800   849.092400    3145.128000      8.78720
std      29.006118    87.798982   41282.539349    116.32032
min       1.000000   607.000000      90.000000      1.00000
25%      24.000000   775.000000     163.000000      1.00000
50%      50.000000   849.000000     284.000000      2.00000
75%      75.000000   925.000000     390.000000      3.00000
max     100.000000   999.000000  704000.000000   2000.00000

In [12]:
# Use the 25% and 75& range of the order_amount to calculate outlier
Q1_Order_Amount_Outlier = 163
Q3_Order_Amount_Outlier = 390
IQR_Order_Amount = Q3_Order_Amount_Outlier - Q1_Order_Amount_Outlier

# Calculate the outlier
Minimum_Outlier = Q1_Order_Amount_Outlier - (1.5*IQR_Order_Amount)
Maximum_Outlier = Q3_Order_Amount_Outlier + (1.5*IQR_Order_Amount)

print('Minimum Outlier: ', Minimum_Outlier)
print('Maximum Outlier: ', Maximum_Outlier)

Minimum Outlier:  -177.5
Maximum Outlier:  730.5


In [13]:
# Only need to use Maximum outlier as a condition since minimum outlier is negative
Outlier_Condition = (df['order_amount'] <= Maximum_Outlier)

# create a df without the outliers
df_no_outlier = df[Outlier_Condition]

# sanity check to see if outliers removed
df_no_outlier.sort_values(by=['order_amount']).tail()

shop_id  user_id  order_amount  total_items payment_method  \
3824       43      736           724            4    credit_card   
2786       43      876           724            4    credit_card   
2312       79      745           724            4          debit   
4659       47      894           725            5           cash   
1124       52      994           730            5    credit_card   

              created_at  
3824 2017-03-25 17:34:51  
2786 2017-03-01 01:41:43  
2312 2017-03-27 09:26:31  
4659 2017-03-01 04:49:59  
1124 2017-03-07 06:54:05

In [14]:
df_no_outlier.describe()

shop_id      user_id  order_amount  total_items
count  4859.000000  4859.000000   4859.000000  4859.000000
mean     49.852645   849.905742    293.715374     1.950196
std      29.049171    86.887496    144.453395     0.919791
min       1.000000   700.000000     90.000000     1.000000
25%      24.000000   776.000000    162.000000     1.000000
50%      50.000000   850.000000    280.000000     2.000000
75%      74.000000   925.000000    380.000000     3.000000
max     100.000000   999.000000    730.000000     5.000000

# Average Order Value of Each Store

In [15]:
# Total number of shops
print('Total number of shops: ', len(df['shop_id'].value_counts()))

Total number of shops:  100


In [16]:
# sort df by shop number for iteration
df = df.sort_values(by=['shop_id'])

# reset index for iteration
df.reset_index(drop=True, inplace=True)

df.head()

shop_id  user_id  order_amount  total_items payment_method  \
0        1      705           474            3          debit   
1        1      934           316            2          debit   
2        1      924           474            3    credit_card   
3        1      951           158            1    credit_card   
4        1      877           474            3           cash   

           created_at  
0 2017-03-13 03:13:36  
1 2017-03-07 03:29:47  
2 2017-03-16 02:31:40  
3 2017-03-04 20:02:50  
4 2017-03-09 17:31:51

In [20]:
# empty lists used for iterative purpose
order_amounts = []
item_counts = []

# lists to be used to create new df
shops = []
AOV = []
avg_num_item_sold = []
sum_sales = []
sum_items = []

# iterate shop id to find AOV and average item_counts
for i, val in enumerate(df['shop_id']):
    if val not in shops:
        shops.append(val)
        # if list is not empty (meaning its not the very 1st iteration)
        if order_amounts:
            # append average and sums to average and sum lists
            AOV.append(sum(order_amounts) / len(order_amounts))
            avg_num_item_sold.append(sum(item_counts) / len(item_counts))
            sum_sales.append(sum(order_amounts))
            sum_items.append(sum(item_counts))
            # clear iterative lists for next shop id
            order_amounts.clear()
            item_counts.clear()
            # add the 1st order amount and item count to next shop id iterative list
            order_amounts.append(df['order_amount'][i])
            item_counts.append(df['total_items'][i])
        else:
            # if list is empty (first iteration) it will add first order amount and item count
            order_amounts.append(df['order_amount'][i])
            item_counts.append(df['total_items'][i])
    else:
        # continue to add order amounts and item counts for the same shop id
        order_amounts.append(df['order_amount'][i])
        item_counts.append(df['total_items'][i])

# append values for the final iteration
AOV.append(sum(order_amounts) / len(order_amounts))
avg_num_item_sold.append(sum(item_counts) / len(item_counts))
sum_sales.append(sum(order_amounts))
sum_items.append(sum(item_counts))

In [26]:
# assert that iteration worked correctly
assert (len(AOV) == len(avg_num_item_sold)) & (len(AOV) == len(sum_items))
assert (len(AOV) == len(sum_sales))

In [29]:
# round numbers in lists
AOV_rounded = [round(num, 2) for num in AOV]
avg_num_item_sold_rounded = [round(num, 2) for num in avg_num_item_sold]

In [35]:
# create store_id list
store_id = [num for num in range(1,101)]
len(store_id)

100

In [46]:
# create dictionary before converting to df
store_data = {'AOV': AOV_rounded, 'avg_num_item_sold': avg_num_item_sold_rounded, 
              'sum_sales': sum_sales, 'sum_items': sum_items}

# create df
df_store = pd.DataFrame(data=store_data, index=store_id)
df_store.index.name = 'store_id'

In [47]:
df_store

AOV  avg_num_item_sold  sum_sales  sum_items
store_id                                                    
1            308.82               1.95      13588         86
2            174.33               1.85       9588        102
3            305.25               2.06      14652         99
4            258.51               2.02      13184        103
5            290.31               2.04      13064         92
6            383.51               2.05      22627        121
7            218.00               1.95      12208        109
8            241.04               1.83      11088         84
9            234.00               1.98      13806        117
10           332.30               2.25      17612        119
11           356.73               1.94      17480         95
12           352.70               1.75      18693         93
13           345.40               2.16      21760        136
14           242.00               2.09      14036        121
15           308.94               2.02      16065        105
16           270.15               1.73      11076         71
17           332.08               1.89      17600        100
18           342.59               2.20      17472        112
19           320.91               1.97      20538        126
20           251.56               1.98      13081        103
21           308.70               2.17      14200        100
22           273.75               1.88      13140         90
23           317.67               2.04      17472        112
24           320.73               2.29      17640        126
25           232.92               1.79      11180         86
26           341.22               1.94      16720         95
27           334.87               1.98      18083        107
28           320.37               1.95      13776         84
29           331.62               2.03      19234        118
30           295.07               1.93      16524        108
31           268.98               2.09      12642         98
32           189.98               1.88       7979         79
33           376.28               2.17      15051         87
34           234.24               1.92      11712         96
35           328.00               2.00      17056        104
36           254.80               1.96      12740         98
37           340.21               2.40      16330        115
38           390.86               2.06      13680         72
39           268.00               2.00      10988         82
40           295.17               1.83      14168         88
41           254.00               2.15      14986        127
42        235101.49             667.90   11990176      34063
43           333.91               1.84      19367        107
44           262.15               1.82      10224         71
45           269.31               1.90      15620        110
46           347.44               2.09      14940         90
47           259.15               1.79      12180         84
48           242.78               2.08       9711         83
49           279.91               2.17      14835        115
50           403.55               2.09      17756         92
51           361.80               1.93      16643         89
52           316.93               2.17      12994         89
53           214.12               1.91      14560        130
54           276.64               2.08      13832        104
55           327.75               1.92      15732         92
56           218.19               1.86       8073         69
57           296.77               2.02      15729        107
58           254.95               1.85      15042        109
59           358.97               2.02      21538        121
60           350.23               1.98      16461         93
61           344.44               2.18      17222        109
62           308.84               1.93      13280         83
63           264.97               1.95      15368        113
64           272.19               2.05      11704  

In [ ]:
        # add a column to df calculating the avg order amount
        order_amounts.append(df['order_amount'][i])
        df['AOV_per_store'][i] = sum(order_amounts) / len(order_amounts)
        # add a column to df calculating the avg total_items sold
        item_counts.append(df['total_items'][i])
        df['avg_num_sold_items'][i] = sum(item_counts) / len(item_counts)


        # add a column to df calculating the avg order amount
        order_amounts.append(df['order_amount'][i])
        df['AOV_per_store'][i] = sum(order_amounts) / len(order_amounts)
        # add a column to df calculating the avg total_items sold
        item_counts.append(df['total_items'][i])
        df['avg_num_sold_items'][i] = sum(item_counts) / len(item_counts)

In [18]:
df.head(100)

shop_id  user_id  order_amount  total_items payment_method  \
0         1      705           474            3          debit   
1         1      934           316            2          debit   
2         1      924           474            3    credit_card   
3         1      951           158            1    credit_card   
4         1      877           474            3           cash   
5         1      725           316            2    credit_card   
6         1      788           316            2    credit_card   
7         1      893           316            2    credit_card   
8         1      786           474            3    credit_card   
9         1      715           316            2          debit   
10        1      886           316            2          debit   
11        1      918           316            2          debit   
12        1      744           316            2           cash   
13        1      991           474            3           cash   
14        1      756           316            2          debit   
15        1      713           316            2           cash   
16        1      907           158            1    credit_card   
17        1      800           316            2    credit_card   
18        1      798           474            3          debit   
19        1      843           316            2          debit   
20        1      847           316            2          debit   
21        1      863           316            2           cash   
22        1      811           316            2          debit   
23        1      734           316            2          debit   
24        1      906           316            2    credit_card   
25        1      836           316            2    credit_card   
26        1      742           158            1    credit_card   
27        1      900           158            1          debit   
28        1      948           632            4           cash   
29        1      929           158            1           cash   
30        1      805           316            2          debit   
31        1      747           158            1          debit   
32        1      887           158            1          debit   
33        1      801           158            1          debit   
34        1      917           316            2          debit   
35        1      847           316            2          debit   
36        1      910           158            1           cash   
37        1      930           474            3           cash   
38        1      797           316            2           cash   
39        1      711           316            2           cash   
40        1      811           474            3    credit_card   
41        1      719           158            1           cash   
42        1      889           158            1          debit   
43        1      738           158            1          debit   
44        2      900           188            2          debit   
45        2      841           188            2           cash   
46        2      977           376            4          debit   
47        2      700           188            2    credit_card   
48        2      941            94            1    credit_card   
49        2      708           188            2    credit_card   
50        2      918            94            1    credit_card   
51        2      704            94            1    credit_card   
52        2      758            94            1    credit_card   
53        2      738           282            3    credit_card   
54        2      831            94            1    credit_card   
55        2      950           282            3          debit   
56        2      978            94            1    credit_card   
57        2      850           282            3           cash   
58        2      843           188            2    credit_card   
59        2      793           470          

# Answer

At first glance, it seems the reason that the average (mean) order amount is so high is that there are some values are outliers (there an order that ordered 2,000 itmes for a total amount of $704,000).

One Solution is to find the median instead of the mean so that it's easier to ignore the outliers. I found the median to be $284, which is a much more believable amount.

Another solution is to find the outlier for the order amount, and create a new dataset with the outliers removed since the outliers are outrageous in amount and are either a mistake or a huge unique type transaction (i.e. warehouse inventory order). After removing the outliers, we found the average (mean) order amount to be ~$294. To note, the median didn't change much with median order amount being $280. Therefore I'm pretty confident using the $294 as the average order amount.

However, since each store only sells one product, it might make sense to calculate the average order value for each store individually.